pkg/proc/gdbserial/gdbserver.go

```diff
@@ -699,41 +689,42 @@ continueLoop:
 	}
 
 	if err := p.updateThreadList(&tu); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	if p.BinInfo().GOOS == "linux" {
 		if err := linutil.ElfUpdateSharedObjects(p); err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 	}
 
-	if err := p.setCurrentBreakpoints(); err != nil {
-		return nil, err
-	}
-
-	for _, thread := range p.threads {
-		if thread.strID == threadID {
-			var err error
-			switch sig {
-			case 0x91:
-				err = errors.New("bad access")
-			case 0x92:
-				err = errors.New("bad instruction")
-			case 0x93:
-				err = errors.New("arithmetic exception")
-			case 0x94:
-				err = errors.New("emulation exception")
-			case 0x95:
-				err = errors.New("software exception")
-			case 0x96:
-				err = errors.New("breakpoint exception")
+	tid, err := strconv.ParseUint(threadID, 16, 32)
+	if thread, ok := p.threads[int(tid)]; err == nil && ok {
+		var err error
+		switch sig {
+		case 0x91:
+			err = errors.New("bad access")
+		case 0x92:
+			err = errors.New("bad instruction")
+		case 0x93:
+			err = errors.New("arithmetic exception")
+		case 0x94:
+			err = errors.New("emulation exception")
+		case 0x95:
+			err = errors.New("software exception")
+		case 0x96:
+			err = errors.New("breakpoint exception")
+		}
+		r := make([]proc.Thread, 0, len(p.threads))
+		for _, t := range p.threads {
```

Comment:
I don't like that we are not using Thread.setbp here and instead just setting a breakpoint on every thread. I think the new logic should follow the old logic. Also the change to the loop looking for the thread doesn't have anything to do with this PR and I'd like to leave it out simply to minimize the patch size.

Reply:
I've added back in `Thread.setbp` in the logic of which threads to return. The change from looping and using the map directly was because the exit condition changed and I thought it clearer to check the map for the thread and then iterate the list of threads instead of having nested loops.